zipline/pipeline/loaders/utils.py

```diff
@@ -133,18 +102,25 @@ def previous_value(date_index, events_by_sid, event_date_field, value_field,
     out = np.full(
         (len(date_index), len(sids)),
         missing_value,
-        dtype=value_field_dtype
+        dtype=field_dtype
     )
     d_n = date_index[-1].asm8
     for col_idx, sid in enumerate(sids):
-        # events_by_sid[sid] is DataFrame mapping knowledge_date to event
-        # date and value.  We don't care about the knowledge date for computing
-        # previous values.
+        # events_by_sid[sid] is a DataFrame mapping knowledge_date to event
+        # date and values.
         df = events_by_sid[sid]
         df = df[df[event_date_field] <= d_n]
+        event_date_vals = df[event_date_field].values
+        # Get knowledge dates corresponding to the values in which we are
+        # interested
+        kd_vals = df[df[event_date_field] <= d_n].index.values
+        # The date at which a previous event is first known is the max of the
+        #  kd and the event date.
+        index_dates = np.array([max([kd, val])
+                                for kd, val in zip(kd_vals, event_date_vals)])
```

Comment:
I described what we want, but that is not what will happen, on t3, the date we report will be `t1`, also, just realized that I failed to properly count 3 elements and did 0, 1, 3 for no reason.